File: Napakalaki/src/napakalaki/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;

import java.util.ArrayList;

/**
 *
 * @author emilio
 */
public class SpecificBadConsequence extends BadConsequence {
    private ArrayList<TreasureKind> specificVisibleTreasures = new ArrayList();
    private ArrayList<TreasureKind> specificHiddenTreasures = new ArrayList();
    
    public SpecificBadConsequence(String text, int levels, ArrayList<TreasureKind> tVisible, ArrayList<TreasureKind> tHidden){
        super(text,levels);
        this.specificVisibleTreasures = tVisible;
        this.specificHiddenTreasures = tHidden;
    }
    
    public ArrayList<TreasureKind> getSpecificVisibleTreasures(){
        return this.specificVisibleTreasures;
    }
    
     public ArrayList<TreasureKind> getSpecificHiddenTreasures(){
        return this.specificHiddenTreasures;
    }
     
    public void substractVisibleTreasure(Treasure t){
        TreasureKind type = t.getType();
        this.specificVisibleTreasures.remove(type);
    }
    
    public void substractHiddenTreasure(Treasure t){
        TreasureKind type = t.getType();
        this.specificVisibleTreasures.remove(type);
    }
    
    
    @Override
    public boolean isEmpty ()
    {
        boolean empty = false;
        
        if ((specificHiddenTreasures.isEmpty()) && 
                (specificVisibleTreasures.isEmpty()))
            empty = true;
        
        return empty;
    }
    
     @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v,
            ArrayList<Treasure> h)
    {
        BadConsequence specificBC;
        
 
        ArrayList<Treasure> copiaV = new ArrayList(v);
        ArrayList<TreasureKind> newSpecificVisible = new ArrayList();

        for (TreasureKind tk : specificVisibleTreasures)
        {
            int i = 0;
            boolean found = false;

            while ((i < copiaV.size()) && !(found))
            {
                if (copiaV.get(i).getType() == tk)
                {
                    found = true;
                    newSpecificVisible.add(tk);
                    copiaV.remove(i);
                }

                i++;
            }
        }

        // FOR HIDDEN TREASURES
        ArrayList<Treasure> copiaH = new ArrayList(h);
        ArrayList<TreasureKind> newSpecificHidden = new ArrayList();

        for (TreasureKind tk : specificHiddenTreasures)
        {
            int i = 0;
            boolean found = false;

            while ((i < copiaH.size()) && !(found))
            {
                if (copiaH.get(i).getType() == tk)
                {
                    found = true;
                    newSpecificHidden.add(tk);
                    copiaH.remove(i);
                }

                i++;
            }
        }

        specificBC = new SpecificBadConsequence(this.getText(),this.getLevels(),
                newSpecificVisible,newSpecificHidden);
        
        return specificBC;
    }
    
    @Override
    public String toString()
    {   
        String out = "";
       
        if (!specificVisibleTreasures.isEmpty()){
        out = "specificVisibleTreasures: ";
         for (int i = 0; i < specificVisibleTreasures.size(); i++)
                out += specificVisibleTreasures.get(i) + " ";
        }
        if (!specificHiddenTreasures.isEmpty()){
            out = "\nspecificHiddenTreasures: ";
             for (int i = 0; i < specificHiddenTreasures.size(); i++)
                    out += specificHiddenTreasures.get(i) + " ";    
       }  
        
       return out;
    }
}

File: NapakalakiRuby/lib/specific_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module NapakalakiGame
class SpecificBadConsequence < BadConsequence
  
  attr_reader :specificVisiblTtreasures # Array of visible treasures
  attr_reader :specificHiddenTreasures  # Array of hidden treasures
  
  def initialize(text, someLevels,
      someSpecificVisibleTreasures, someSpecificHiddenTreasures)
    super(text, someLevels)
    @specificVisibleTreasures = someSpecificVisibleTreasures
    @specificHiddenTreasures = someSpecificHiddenTreasures
  end
  
  def substract_visible_treasure (t)
    if (!@specificVisibleTreasures.empty?)
      @specificVisibleTreasures.delete(t.type)
    end

  end
  
   def substract_hidden_treasure (t)
    if(!@specificHiddenTreasures.empty?)
            @specificHiddenTreasures.delete(t.type)
    end
    
  end
  
   def is_empty ()
    result = false
    if ((@specificVisibleTreasures.empty?) && (@specificHiddenTreasures.empty?))
      result = true
    end   
    result
  end
  
  def adjust_to_fit_treasure_lists(v, h)
  
      # CASE 2:
      copiaV = Array.new(v)
      vTreasures = Array.new()
      
      for t in @specificVisibleTreasures
        i = 0;
        found = false;
        while (i < copiaV.size) && !(found)
          if (copiaV[i].type == t)
             found = true
             vTreasures << t
       
             #copia_v.delete_at(copia_v.index(i) || copia_v.length)
             copiaV.delete(v[i])
          end     
          i = i+1
        end
      end
                  
      copiaH = Array.new(h)
      hTreasures = Array.new()

      for t in @specificHiddenTreasures
         i = 0
         found = false
         while (i < copiaH.size) && !(found)
            
             if (copiaH[i].type == t)
                 found = true;
                 hTreasures << t
                #copia_h.delete_at(copia_h.index(i) || copia_h.length)
                copiaH.delete(h[i])
             end   
             i = i+1
         end
      end
      adjustBC = SpecificBadConsequence.new(@text,@levels,vTreasures,hTreasures)
      return adjustBC
  end
  
   def to_s
    
    respuesta = super.to_s
    respuesta += " \nvTreasures: #{@specificVisibleTreasures}\n+ hTreasures #{@specificHiddenTreasures} "
 
  end
  
  
end
end